File: bugsnag-android-core/src/main/java/com/bugsnag/android/SessionStore.java
package com.bugsnag.android;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.File;
import java.util.Comparator;
import java.util.Locale;
import java.util.UUID;

/**
 * Store and flush Sessions which couldn't be sent immediately due to
 * lack of network connectivity.
 */
class SessionStore extends FileStore {

    static final Comparator<File> SESSION_COMPARATOR = new Comparator<File>() {
        @Override
        public int compare(File lhs, File rhs) {
            if (lhs == null && rhs == null) {
                return 0;
            }
            if (lhs == null) {
                return 1;
            }
            if (rhs == null) {
                return -1;
            }
            String lhsName = lhs.getName();
            String rhsName = rhs.getName();
            return lhsName.compareTo(rhsName);
        }
    };

    SessionStore(@NonNull Context appContext,
                 @NonNull ImmutableConfig config,
                 @NonNull Logger logger,
                 @Nullable Delegate delegate) {
        super(appContext, "/bugsnag-sessions/", config.getMaxPersistedSessions(),
                SESSION_COMPARATOR, logger, delegate);
    }

    @NonNull
    @Override
    String getFilename(Object object) {
        return String.format(Locale.US, "%s%s%d_v2.json",
            storeDirectory, UUID.randomUUID().toString(), System.currentTimeMillis());
    }

}

File: bugsnag-android-core/src/main/java/com/bugsnag/android/SessionStore.kt
package com.bugsnag.android

import com.bugsnag.android.SessionFilenameInfo.Companion.defaultFilename
import com.bugsnag.android.SessionFilenameInfo.Companion.findTimestampInFilename
import com.bugsnag.android.internal.ImmutableConfig
import java.io.File
import java.util.Calendar
import java.util.Comparator
import java.util.Date

/**
 * Store and flush Sessions which couldn't be sent immediately due to
 * lack of network connectivity.
 */
internal class SessionStore(
    private val config: ImmutableConfig,
    logger: Logger,
    delegate: Delegate?
) : FileStore(
    File(
        config.persistenceDirectory.value, "bugsnag/sessions"
    ),
    config.maxPersistedSessions,
    SESSION_COMPARATOR,
    logger,
    delegate
) {
    fun isTooOld(file: File?): Boolean {
        val cal = Calendar.getInstance()
        cal.add(Calendar.DATE, -60)
        return findTimestampInFilename(file!!) < cal.timeInMillis
    }

    fun getCreationDate(file: File?): Date {
        return Date(findTimestampInFilename(file!!))
    }

    companion object {
        val SESSION_COMPARATOR: Comparator<in File?> = Comparator { lhs, rhs ->
            if (lhs == null && rhs == null) {
                return@Comparator 0
            }
            if (lhs == null) {
                return@Comparator 1
            }
            if (rhs == null) {
                return@Comparator -1
            }
            val lhsName = lhs.name
            val rhsName = rhs.name
            lhsName.compareTo(rhsName)
        }
    }

    override fun getFilename(obj: Any?): String {
        val sessionInfo = defaultFilename(obj, config)
        return sessionInfo.encode()
    }
}
